pkg/api/routes/components.go

```diff
@@ -47,3 +47,60 @@ func fetchServerComponents(client internalfleetdb.Client, srvid uuid.UUID, l *za
 	}
 	return comps, nil
 }
+
+func verifyComponent(c internalfleetdb.Client, server *rivets.Server, dev *rivets.Server, log *zap.Logger) {
+	alloyComponentsMap := make(map[string][]*rivets.Component)
+	for _, component := range dev.Components {
+		slug := component.Name
+		if _, ok := alloyComponentsMap[slug]; !ok {
+			alloyComponentsMap[slug] = make([]*rivets.Component, 0)
+		}
+		alloyComponentsMap[slug] = append(alloyComponentsMap[slug], component)
+	}
+
+	for _, fleetComponent := range server.Components {
+		slug := fleetComponent.Name
+		alloyComponents, ok := alloyComponentsMap[slug]
+		if !ok {
+			// no slug in alloy list, fleet component not in alloy list
+			fields := []zap.Field{
+				zap.String("device.id", server.ID),
+				zap.String("component", slug),
+				zap.String("expected(alloy)", "nil"),
+				zap.String("current(fleetdb)", fleetComponent.Firmware.Installed),
+			}
+			log.Warn("fleetdb component not listed in alloy", fields...)
+			continue
+		}
+
+		var listedByAlloy bool
+		for _, alloyComponent := range alloyComponents {
+			if alloyComponent.Vendor+alloyComponent.Model == fleetComponent.Vendor+fleetComponent.Model {
```

Comment:
🤔 We're checking equality of two strings. Concatenating them and turning this into one test is efficient but potentially harder to understand for a maintainer. OK for now.

Reply:
Agree to make it a var once we have multiple use cases.